In [1]:
import pandas as pd

In [2]:
INPUT_FILE_1 = '../data-scraping/product-pages/data/sample-scraped/flipkart_products_scraped_part_1.csv'
INPUT_FILE_2 = '../data-scraping/product-pages/data/sample-scraped/flipkart_products_scraped_part_2.csv'

OUTPUT_FILE = '../dataset/flipkart_products_merged.csv'

In [3]:
df_1 = pd.read_csv(INPUT_FILE_1)
df_2 = pd.read_csv(INPUT_FILE_2)

In [4]:
merged = pd.concat([df_1, df_2], ignore_index=True)

In [5]:
merged.to_csv(OUTPUT_FILE, index=False)

In [6]:
df_merged = pd.read_csv(OUTPUT_FILE)

In [7]:
df_merged.head()

url  \
0  https://www.flipkart.com/homestic-4-5-l-plasti...   
1  https://www.flipkart.com/dreamlivproducts-gree...   
2  https://www.flipkart.com/beautum-born-winner-j...   
3  https://www.flipkart.com/keycent-back-cover-ap...   
4  https://www.flipkart.com/voddmars-car-fan-12v-...   

                                               title   price discount  \
0                   HOMESTIC 4.5 L Plastic Water Jug  ₹1,999  76% off   
1  DreamlivproductsBhaiya Bhabhi Attractive rakhi...    ₹499  58% off   
2  Beautum BORN WINNER Jermel Name Printed Cerami...    ₹999  74% off   
3  KEYCENT Back Cover for APPLE iPhone 15 Plus KR...  ₹1,499  80% off   
4  Voddmars Car Fan 12V 360 Head 2 Speed Quiet St...  ₹1,999  77% off   

  discounted_price  rating_value rating_counts review_counts  \
0             ₹473           3.0           NaN           NaN   
1             ₹207           4.0           NaN           NaN   
2             ₹258           4.4           NaN           NaN   
3             ₹297           4.1           NaN           NaN   
4             ₹453           3.4           NaN           NaN   

                                           hierarchy  \
0  Home > Kitchen, Cookware & Serveware > Tablewa...   
1  Home > Festive Decor & Gifting > Festive Gifts...   
2  Home > Kitchen, Cookware & Serveware > Tablewa...   
3  Home > Mobiles & Accessories > Mobile Accessor...   
4  Home > Automotive Accessories > Vehicle Utilit...   

                                         description  
0            3★5 Ratings &0 Reviews: 5★4★3★2★1★12002  
1  Sales Package: 5 RAKHI SET 1 Greeting Card 1 R...  
2  We shall not be liable for the cheap imitated ...  
3  APPLE iPhone 15 Plus Back Cover by KEYCENT Sty...  
4                                                NaN

In [8]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 520911 entries, 0 to 520910
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   url               520911 non-null  object 
 1   title             520851 non-null  object 
 2   price             508728 non-null  object 
 3   discount          508531 non-null  object 
 4   discounted_price  509775 non-null  object 
 5   rating_value      476269 non-null  float64
 6   rating_counts     22661 non-null   object 
 7   review_counts     22661 non-null   object 
 8   hierarchy         505120 non-null  object 
 9   description       439357 non-null  object 
dtypes: float64(1), object(9)
memory usage: 39.7+ MB


In [ ]:
df_merged = df_merged.drop(columns=['rating_counts', 'review_counts', 'discounted_price'])

In [ ]:
df_merged['price'] = (
    df_merged['price']
    .str.replace('₹', '', regex=False)
    .str.replace(',', '', regex=False)
    .astype(float)
)

In [20]:
df_merged['discount'] = (
    df_merged['discount']
    .str.replace('% off', '', regex=False)
    .astype(float)
)

In [24]:
df_merged.isnull().sum()

url                 0
title              60
price           12183
discount        12380
rating_value    44642
hierarchy       15791
description     81554
dtype: int64

In [25]:
df_merged = df_merged.dropna(subset=['price', 'discount'])

In [27]:
df_merged = df_merged.dropna(subset=['title'])

In [29]:
df_merged['description'].fillna('No Description', inplace=True)


/var/folders/7p/_l5kmmts4m3886n2ycnm87380000gn/T/ipykernel_2992/2745207510.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_merged['description'].fillna('No Description', inplace=True)


In [31]:
df_merged['hierarchy'].fillna('Unknown', inplace=True)

/var/folders/7p/_l5kmmts4m3886n2ycnm87380000gn/T/ipykernel_2992/3267907259.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_merged['hierarchy'].fillna('Unknown', inplace=True)


In [32]:
df_merged.isnull().sum()

url                 0
title               0
price               0
discount            0
rating_value    34028
hierarchy           0
description         0
dtype: int64

In [34]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 508525 entries, 0 to 520910
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   url           508525 non-null  object 
 1   title         508525 non-null  object 
 2   price         508525 non-null  float64
 3   discount      508525 non-null  float64
 4   rating_value  474497 non-null  float64
 5   hierarchy     508525 non-null  object 
 6   description   508525 non-null  object 
dtypes: float64(3), object(4)
memory usage: 31.0+ MB


In [43]:
df_merged["hierarchy_list"] = df_merged["hierarchy"].apply(
    lambda x: [p.strip() for p in x.split(">")] if pd.notna(x) else None
)

In [44]:
df_merged["hierarchy_depth"] = df_merged["hierarchy_list"].apply(lambda x: len(x) if x else 0)

In [46]:
df_merged["hierarchy_depth"].describe()

count    508525.000000
mean          6.557750
std           0.804129
min           1.000000
25%           6.000000
50%           7.000000
75%           7.000000
max           8.000000
Name: hierarchy_depth, dtype: float64

In [47]:
max_depth = 8

In [ ]:
for i in range(max_depth):
    df_merged[f"level_{i+1}"] = df_merged["hierarchy_list"].apply(
        lambda x: x[i] if x and len(x) > i else None
    )

In [54]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 508525 entries, 0 to 520910
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   url              508525 non-null  object 
 1   title            508525 non-null  object 
 2   price            508525 non-null  float64
 3   discount         508525 non-null  float64
 4   rating_value     474497 non-null  float64
 5   hierarchy        508525 non-null  object 
 6   description      508525 non-null  object 
 7   hierarchy_list   508525 non-null  object 
 8   hierarchy_depth  508525 non-null  int64  
 9   level_1          508525 non-null  object 
 10  level_2          503849 non-null  object 
 11  level_3          503849 non-null  object 
 12  level_4          503651 non-null  object 
 13  level_5          498130 non-null  object 
 14  level_6          486644 non-null  object 
 15  level_7          327815 non-null  object 
 16  level_8          2317 non-null    object 
d

```
non_list_cols = [c for c in df_merged.columns 
                  if df_merged[c].apply(lambda x: not isinstance(x, list)).all()]

df_merged[non_list_cols].nunique()
```

In [55]:
import re

In [56]:
emoji_pattern = re.compile(
    "[" 
    "\U0001F600-\U0001F64F"  # emoticons
    "\U0001F300-\U0001F5FF"  # symbols & pictographs
    "\U0001F680-\U0001F6FF"  # transport
    "\U0001F1E0-\U0001F1FF"  # flags
    "\U00002700-\U000027BF"  # dingbats
    "\U0001F900-\U0001F9FF"  # supplemental symbols
    "\U0001FA70-\U0001FAFF"  # extended symbols
    "\U00002600-\U000026FF"  # misc symbols
    "\U00002B00-\U00002BFF"  # arrows
    "]+",  
    flags=re.UNICODE
)

In [57]:
def remove_emojis(text):
    if pd.isna(text):
        return text
    return emoji_pattern.sub('', text)

In [58]:
df_merged["title"] = df_merged["title"].apply(remove_emojis)
df_merged["description"] = df_merged["description"].apply(remove_emojis)

In [59]:
df_merged.head()

url  \
0  https://www.flipkart.com/homestic-4-5-l-plasti...   
1  https://www.flipkart.com/dreamlivproducts-gree...   
2  https://www.flipkart.com/beautum-born-winner-j...   
3  https://www.flipkart.com/keycent-back-cover-ap...   
4  https://www.flipkart.com/voddmars-car-fan-12v-...   

                                               title   price  discount  \
0                   HOMESTIC 4.5 L Plastic Water Jug  1999.0      76.0   
1  DreamlivproductsBhaiya Bhabhi Attractive rakhi...   499.0      58.0   
2  Beautum BORN WINNER Jermel Name Printed Cerami...   999.0      74.0   
3  KEYCENT Back Cover for APPLE iPhone 15 Plus KR...  1499.0      80.0   
4  Voddmars Car Fan 12V 360 Head 2 Speed Quiet St...  1999.0      77.0   

   rating_value                                          hierarchy  \
0           3.0  Home > Kitchen, Cookware & Serveware > Tablewa...   
1           4.0  Home > Festive Decor & Gifting > Festive Gifts...   
2           4.4  Home > Kitchen, Cookware & Serveware > Tablewa...   
3           4.1  Home > Mobiles & Accessories > Mobile Accessor...   
4           3.4  Home > Automotive Accessories > Vehicle Utilit...   

                                         description  \
0                  35 Ratings &0 Reviews: 5432112002   
1  Sales Package: 5 RAKHI SET 1 Greeting Card 1 R...   
2  We shall not be liable for the cheap imitated ...   
3  APPLE iPhone 15 Plus Back Cover by KEYCENT Sty...   
4                                     No Description   

                                      hierarchy_list  hierarchy_depth level_1  \
0  [Home, Kitchen, Cookware & Serveware, Tablewar...                6    Home   
1  [Home, Festive Decor & Gifting, Festive Gifts,...                6    Home   
2  [Home, Kitchen, Cookware & Serveware, Tablewar...                6    Home   
3  [Home, Mobiles & Accessories, Mobile Accessori...                7    Home   
4  [Home, Automotive Accessories, Vehicle Utility...                6    Home   

                         level_2                      level_3  \
0  Kitchen, Cookware & Serveware       Tableware & Dinnerware   
1        Festive Decor & Gifting                Festive Gifts   
2  Kitchen, Cookware & Serveware       Tableware & Dinnerware   
3          Mobiles & Accessories           Mobile Accessories   
4         Automotive Accessories  Vehicle Utility Accessories   

              level_4                              level_5  \
0                Jugs                        HOMESTIC Jugs   
1  Rakhi & Rakhi Sets  Dreamlivproducts Rakhi & Rakhi Sets   
2                Mugs                         Beautum Mugs   
3      Cases & Covers              Designer Cases & Covers   
4   Car Interior Fans           Voddmars Car Interior Fans   

                                             level_6  \
0                   HOMESTIC 4.5 L Plastic Water Jug   
1  Dreamlivproducts Bhaiya Bhabhi Attractive rakh...   
2  Beautum BORN WINNER Jermel Name Printed Cerami...   
3                    KEYCENT Designer Cases & Covers   
4  Voddmars Car Fan 12V 360 Head 2 Speed Quiet St...   

                                             level_7 level_8  
0                                               None    None  
1                                               None    None  
2                                               None    None  
3  KEYCENT Back Cover for APPLE iPhone 15 Plus KR...    None  
4                                               None    None

In [61]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 508525 entries, 0 to 520910
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   url              508525 non-null  object 
 1   title            508525 non-null  object 
 2   price            508525 non-null  float64
 3   discount         508525 non-null  float64
 4   rating_value     474497 non-null  float64
 5   hierarchy        508525 non-null  object 
 6   description      508525 non-null  object 
 7   hierarchy_list   508525 non-null  object 
 8   hierarchy_depth  508525 non-null  int64  
 9   level_1          508525 non-null  object 
 10  level_2          503849 non-null  object 
 11  level_3          503849 non-null  object 
 12  level_4          503651 non-null  object 
 13  level_5          498130 non-null  object 
 14  level_6          486644 non-null  object 
 15  level_7          327815 non-null  object 
 16  level_8          2317 non-null    object 
d

**508525 entries**

| #   | Column          | Non-Null Count | Dtype    |
|-----|------------------|----------------|----------|
| 0   | url              | 508,525        | object   |
| 1   | title            | 508,525        | object   |
| 2   | price            | 508,525        | float64  |
| 3   | discount         | 508,525        | float64  |
| 4   | rating_value     | 474,497        | float64  |
| 5   | hierarchy        | 508,525        | object   |
| 6   | description      | 508,525        | object   |
| 7   | hierarchy_list   | 508,525        | object   |
| 8   | hierarchy_depth  | 508,525        | int64    |
| 9   | level_1          | 508,525        | object   |
| 10  | level_2          | 503,849        | object   |
| 11  | level_3          | 503,849        | object   |
| 12  | level_4          | 503,651        | object   |
| 13  | level_5          | 498,130        | object   |
| 14  | level_6          | 486,644        | object   |
| 15  | level_7          | 327,815        | object   |
| 16  | level_8          | 2,317          | object   |


In [62]:
df_merged.to_csv('../dataset/flipkart_products_preprocessed_1.csv')